utils/controller_debug.py

```diff
@@ -0,0 +1,1058 @@
+            That way any old (possibly incomplete) data that had previously
+            been printed by the firmware to the print buffer will be removed
+            before the display starts
+
+<filename>  If a file name is given, the data received will be
+            written to the file as well as displayed.
+"""
+
+        cl = line.split()
+
+        # Optionally flush the buffer before we start displaying
+        if "--flush" in cl:
+            while len(SendCmd(OP_PBREAD)) > 0:
+                pass
+            cl.remove("--flush")
+
+        if len(cl) > 0:
+            fp = open(cl[0], "w")
+        else:
+            fp = None
+
+        try:
+            while True:
+                dat = SendCmd(OP_PBREAD)
+                if len(dat) < 1:
+                    continue
+
+                S = "".join([chr(x) for x in dat])
+
+                sys.stdout.write(S)
+                sys.stdout.flush()
+
+                if fp != None:
+                    fp.write(S)
+
+        except KeyboardInterrupt:
+            print()
+            ReSync()
+            pass
+
+    def do_EOF(self, line):
+        return True
+
+    def do_exit(self, line):
+        return True
+
+    def do_get(self, line):
+        cl = line.split()
+        if len(cl) < 1:
+            print("Please give the variable name to read")
+            return
+
+        print(GetVar(cl[0]))
+
+    def complete_get(self, text, line, begidx, endidx):
+        var = text
+
+        out = []
+        for i in varDict.keys():
+            if i.startswith(var):
+                out.append(i)
+
+        return out
+
+    def help_get(self):
+        global varDict
+        print("Read the value of a debug variable and display it\n")
+        print("Variables currently defined:")
+        for k in varDict.keys():
+            print("   %-10s - %s" % (k, varDict[k].help))
+
+    def do_set(self, line):
+        cl = line.split()
+        if len(cl) < 2:
+            print("Please give the variable name and value")
+            return
+        SetVar(cl[0], cl[1])
+
+    def complete_set(self, text, line, begidx, endidx):
+        var = text
+
+        out = []
+        for i in varDict.keys():
+            if i.startswith(var):
+                out.append(i)
+
+        return out
+
+    def help_set(self):
+        global varDict
+        print("Set the value of a debug variable\n")
+        print("Variables currently defined:")
+        for k in varDict.keys():
+            print("   %-10s - %s" % (k, varDict[k].help))
+
+    def do_trace(self, line):
+        """
+       This command is used to access the trace buffer on the controller.
+       A sub-command must be passed as an option:
+
+       trace flush
+         Stops the trace if one was on-going and flushes the trace buffer
+
+       trace start
+         Starts the trace collecting data.
+
+       trace graph
+         This will download the data and display it graphically
```

Comment:
Just the current contents or interactively?

Reply:
Just the current contents of the buffer.  Typically you set up and start a trace, wait for the buffer to fill, then download and either save to file or graph it. 
The Python code could read out while the trace is being collected to it's buffer, but you wouldn't be able to sample as quickly because the data can be added to the buffer much faster then the serial port can read it out.